Fixes #36547 - Fix parsing of Ubuntu version in fact parsers

goarsna opened the pull request.

goarsna committed:
Fixes #36547 - Fix parsing of Ubuntu version in fact parsers

ekohl commented on app/services/puppet_fact_parser.rb:
```diff
@@ -5,7 +5,13 @@ def operatingsystem
 
     if orel.present?
-      major, minor = orel.split('.', 2)
+      version_splits = orel.split('.')
+      if os_name == 'ubuntu'
+        major = "#{version_splits[0]}.#{version_splits[1]}"
+        minor = version_splits[2] if version_splits.length == 3
+      else
+        major, minor = version_splits
+      end
       major = major.to_s.gsub(/\D/, '')
       minor = minor.to_s.gsub(/[^\d\.]/, '')
```
I've been thinking about trusting the Facter reported major version instead. There it reports `22.04` as the major, instead of 22 as major and 04 as minor.

ekohl reviewed (commented):
While technically correct, I think this may cause problems. For example, in searching it splits on the dot here:

https://github.com/ekohl/foreman/blob/0eb81758b02d0a4a397e14e076fec6a70a3aa2ee/app/models/operatingsystem.rb#L169-L178

I also wonder about migrations. Should we introduce a DB migration where we change the existing OS versions in the DB?

goarsna commented:
Hey @ekohl 
Thanks for your feedback. And I have to apologize that my response took so long...

I have incorporated your feedback. I also searched for all areas of code where the version is parsed / used and fixed them. Hopefully I got them all. :)

Regarding db migrations: Should we consider anything else despite changing the version of the operating system (for example migrating to existing OSes with the correct version)?

goarsna commented:
Hey @ekohl,
I have rebased my PR and added a db migration that migrates Ubuntu OSes with non-empty minor version if no Ubuntu OS with the target major version is present.

Please let me know  if anything is missing.

ekohl commented on app/models/operatingsystem.rb:
```diff
@@ -170,7 +170,11 @@ def self.find_by_to_label(str)
     return os if os
     a = str.split(" ")
-    b = a[1].split('.') if a[1]
+    if a[0] =~ /ubuntu/i
```
`=~` has some changes in Ruby 3.2 and a better alternative is to use `.match?` when you can. Also has a better performance, so newer versions of RuboCop also flag this
```suggestion
    if a[0]&.match?(/ubuntu/i)
```

goarsna replied:
I switched to using the normalized OS name as proposed below.

ekohl commented on app/models/operatingsystem.rb:
```diff
@@ -170,7 +170,11 @@ def self.find_by_to_label(str)
     a = str.split(" ")
-    b = a[1].split('.') if a[1]
+    if a[0] =~ /ubuntu/i
+      b = a[1], nil if a[1]
```
Some tests for this would be nice. I'd be worried about `22.04.2` because Ubuntu does have minor releases.

goarsna replied:
Ok, I think I was a little bit to hasty with my changes... Indeed at place if there is a minor version present in the string, I should take it into account :sweat_smile:

ekohl commented on app/services/foreman_ansible/operating_system_parser.rb:
```diff
@@ -68,8 +70,12 @@ def os_release
     def os_minor
-      _, minor = os_release&.split('.', 2) ||
+      if os_name == 'Ubuntu'
+        minor = nil
```
Ubuntu does have minor releases. There's `22.04.2` now for example. I don't know if this matters in things like URLs to repositories. You do track it for Puppet so things may flip-flop.

ekohl commented on app/services/katello/rhsm_fact_parser.rb:
```diff
@@ -55,7 +55,13 @@ def operatingsystem
       os_name = distribution_to_puppet_os(name)
-      major, minor = version.split('.')
+      version_splits = version.split('.')
+      if name =~ /ubuntu/i
```
Same comment about `.match?` as above. Or we already have a normalized name so `if os_name == 'Ubuntu'` should also be reliable.

ekohl commented on app/services/katello/rhsm_fact_parser.rb:
```diff
@@ -55,7 +55,13 @@ def operatingsystem
 
       os_name = distribution_to_puppet_os(name)
-      major, minor = version.split('.')
+      version_splits = version.split('.')
```
It should be noted you can pass a parameter to `.split` about how often it is allowed to split. So:
```
[1] pry(main)> '22.04.2'.split('.', 3)
=> ["22", "04", "2"]
[2] pry(main)> '7.8'.split('.', 3)
=> ["7", "8"]
```

Perhaps it should be:
```ruby
if name.match?(/ubuntu/i)
  x, y, minor = version.split('.', 3)
  major = "#{x}.#{y}"
else
  major, minor = version.split('.', 2)
end
```

goarsna replied:
Thanks for the improvement. I also now rely on the normalized OS name and applied your suggestion

ekohl commented on app/services/puppet_fact_parser.rb:
```diff
@@ -5,9 +5,14 @@ def operatingsystem
-      major = major.to_s.gsub(/\D/, '')
-      minor = minor.to_s.gsub(/[^\d\.]/, '')
+      if os_name =~ /ubuntu/i
+        major = os_major_version
+        minor = os_minor_version
```
I've been debating making this the behavior everywhere and would be OK with trying that.

goarsna replied:
So that we simply leave out the if / else part and use `os_major_version` and `os_minor_version` for all operating systems? I also noticed that the `if orel.present?` isn't necessary here anymore, as the whole part from line 8 to line 24 is no longer dependent on it.

ekohl replied:
indeed, get rid of orel altogether.

goarsna replied:
I had a look at relying on the `major` and `minor` versions reported by facter and it doesn't fit for all operating systems. The `os_release` often uses the `full` release reported by facter and parses it as necessary. Relying on the reported `major` and `minor` versions breaks some of the OSes (like Solaris and CentOS 8).

IMO it would be best to rework detection of the OS version in the PuppetFactParser in a separate PR.

ekohl commented on db/migrate/20230719080900_fix_ubuntu_versions.rb:
```diff
@@ -0,0 +1,13 @@
+    #   * If not, we update the version
+    Operatingsystem.where('name ~* :name', name: 'ubuntu')
+                    .merge(Operatingsystem.where.not(minor: [nil, ''])).each do |os|
+      if !Operatingsystem.where('name ~* :name', name: 'ubuntu').merge(Operatingsystem.where(major: "#{os.major}.#{os.minor}")).merge(Operatingsystem.where(minor: [nil, ''])).first
```
I'm not familiar with `.merge` and the ORM. Does that translate to something in SQL or is it doing 3 queries for every OS?

goarsna replied:
The merge method can be used to combine two Active Record relations into one. I used it to combine the case insensitive regex search with the other parameters. But I just noticed I could have combined the two calls to `merge` into one.

But if I can rely on the normalized OS name "Ubuntu" both queries in line 6/7 and line 8 get really simple. I adjusted them accordingly.

goarsna committed:
Refs #36547 - Fix parsing of Ubuntu version in fact parsers

goarsna committed:
Refs #36547 - Fix parsing of Ubuntu version in fact parsers

goarsna committed:
Refs #36547 - Fix parsing of Ubuntu version in fact parsers

goarsna committed:
Refs #36547 - Fix parsing of Ubuntu version in fact parsers

ekohl commented on db/migrate/20230719080900_fix_ubuntu_versions.rb:
```diff
@@ -0,0 +1,12 @@
+    #   * Check if there is already an operating system present with the same major and minor version combined in the major field
+    #   * If not, we update the version
+    Operatingsystem.where(name: 'Ubuntu').where.not(minor: [nil, '']).each do |os|
+      unless Operatingsystem.find_by(name: 'Ubuntu', major: "#{os.major}.#{os.minor}", minor: [nil, ''])
```
In general it's a bad practice to run an SQL query for every item, which results in 1+n queries. Here it may not be too bad because it's likely a small result set and a migration only runs once.

goarsna replied:
I converted the db migration to a single joined update query and tested the behavior with following content in the `operatingsystems` table:

```
 id | major |    name     | minor |        id | major |    name     | minor |
----+-------+-------------+-------+       ----+-------+-------------+-------+
  1 | 18    | Ubuntu      | 04    |         1 | 18.04 | Ubuntu      |       |
  2 | 20.04 | Ubuntu      |       |   =>    2 | 20.04 | Ubuntu      |       |
  3 | 22.04 | Ubuntu      | 3     |         3 | 22.04 | Ubuntu      | 3     |
  4 | 20    | Ubuntu      | 04    |         4 | 20    | Ubuntu      | 04    |
```

ekohl commented on app/models/operatingsystem.rb:
```diff
@@ -170,7 +170,14 @@ def self.find_by_to_label(str)
     return os if os
     a = str.split(" ")
-    b = a[1].split('.') if a[1]
+    if a[0] == 'Ubuntu'
```
This is getting hard to read, so I started to rewrite it:
```ruby
name, version = str.split(" ")
major = minor = nil
if name == 'Ubuntu'
  if version
    x, y, minor = version.split('.', 3)
    major = "#{x}.#{y}"
  end
else
  major, minor = version.split('.') if version
end

cond = {:name => name}
cond[:major] = major if major
cond[:minor] = minor if minor

find_by(cond)
```

Then it made me question the logic. So how about:
```ruby
name, version = str.split(" ")

cond = {:name => name}

if version
  if name == 'Ubuntu'
    x, y, minor = version.split('.', 3)
    major = "#{x}.#{y}"
  else
    major, minor = version.split('.')
  end

  cond[:major] = major if major
  cond[:minor] = minor if minor
end

find_by(cond)
```

But this function needs some tests to cover this.

goarsna replied:
Thanks, @ekohl, I've changed the method as suggested and added tests for Debian and Ubuntu. In the course of this I added Debian 12 to the OS fixtures and updated the Ubuntu 10.10 OS. Or is there a specific reason this old version is used?

goarsna committed:
Refs #36547 - Fix parsing of Ubuntu version in fact parsers

goarsna committed:
Refs #36547 - Fix parsing of Ubuntu version in fact parsers

goarsna committed:
Refs #36547 - Fix parsing of Ubuntu version in fact parsers

goarsna committed:
Refs #36547 - Fix parsing of Ubuntu version in fact parsers

goarsna commented:
Good morning and a happy new year :)
I finally made it to work on this PR again and fixed some things, added tests and fixed failing tests.
I'm looking forward to any feedback :)

goarsna committed:
Refs #36547 - Fix parsing of Ubuntu version in fact parsers

ekohl commented on test/fixtures/operatingsystems.yml:
```diff
@@ -18,15 +18,25 @@ redhat:
+  architectures: x86_64
+  title: 'Debian 12'
+
+ubuntu2204:
```
We should avoid using fixtures and use factories instead.

ekohl commented on test/models/operatingsystem_test.rb:
```diff
@@ -56,6 +56,18 @@ class OperatingsystemTest < ActiveSupport::TestCase
   end
 
+  test "should find Debian by fullname string" do
+    str = "Debian 12"
+    os = Operatingsystem.find_by_to_label(str)
+    assert_equal str, os.fullname
+  end
+
+  test "should find Ubuntu by fullname string" do
+    str = "Ubuntu 22.04.3"
+    os = Operatingsystem.find_by_to_label(str)
+    assert_equal str, os.fullname
+  end
```
Please use factories. Define if needed in `test/factories/operatingsystem.rb` but given how specific this is to test parsing of version numbers I think defining it locally makes sense as well.

ekohl commented on test/static_fixtures/facts/puppet_ubuntu_22_04_3_facter_2.5.json:
We use [facterdb](https://github.com/voxpupuli/facterdb) to store these facts to avoid having these huge fixtures. If needed, raise the minimum version to one that includes Ubuntu 22.04 facts:
https://github.com/theforeman/foreman/blob/75017ef201b8e0de64559f4613e0ce8e8fdc34fb/bundler.d/test.rb#L12

I'd argue the existing files that we have should also be dropped and replaced with reading them from facterdb.

goarsna replied:
I agree, but replacing existing static fixtures should be handled separately.

goarsna commented:
Thanks @ekohl , I'll have a look at the usage of factories and facterdb and adjust my changes.
